src/test/java/com/arpnetworking/commons/math/AccumulatorTest.java

```diff
@@ -0,0 +1,215 @@
+        final double max = 1.0e16;
+        final double min = -1.0e16;
+        for (int i = 0; i < 4000000; ++i) {
+            valueDataSet.add(valueRandom.nextDouble() * (max - min) + min);
+        }
+        RANDOM_VALUE_DATA_SET = Collections.unmodifiableList(valueDataSet);
+
+        // Random cancellations data set
+        final List<Double> cancelDataSet = new ArrayList<>();
+        final Random cancelRandom = new Random(1591215254);
+        for (int i = 0; i < 4000000; ++i) {
+            final int group = cancelRandom .nextInt(4);
+            switch (group) {
+                case 0:
+                    cancelDataSet.add(1.0);
+                    break;
+                case 1:
+                    cancelDataSet.add(1.0e16);
+                    break;
+                case 2:
+                    cancelDataSet.add(-1.0);
+                    break;
+                case 3:
+                    cancelDataSet.add(-1.0e16);
+                    break;
+                default:
+                    throw new IllegalStateException("This should never happen");
+            }
+        }
+        RANDOM_CANCEL_DATA_SET = Collections.unmodifiableList(cancelDataSet);
+    }
+
+    private final Supplier<Accumulator> _accumulatorSupplier;
+    private final double _differentMagnitudesMargin;
+    private final double _differentSignsMargin;
+    private final double _randomCancellationMargin;
+    private final double _randomValuesMargin;
+
+    public AccumulatorTest(
+            final Supplier<Accumulator> accumulatorSupplier,
+            final Double differentMagnitudesMargin,
+            final Double differentSignsMargin,
+            final Double randomCancellationMargin,
+            final Double randomValuesMargin) {
+        _accumulatorSupplier = accumulatorSupplier;
+        _differentMagnitudesMargin = differentMagnitudesMargin;
+        _differentSignsMargin = differentSignsMargin;
+        _randomCancellationMargin = randomCancellationMargin;
+        _randomValuesMargin = randomValuesMargin;
+    }
+
+    @Parameterized.Parameters(name = "{index}: {0}")
+    public static Collection<Object[]> parameters() {
+        return Arrays.asList(
+                new Object[]{
+                        (Supplier<Accumulator>) NaiveAccumulator::new,
+                        1.0E10,     // magnitudes
+                        0.0001,     // signs
+                        1.1e3,      // random cancel
+                        3.9e6},     // random values
+                new Object[]{
+                        (Supplier<Accumulator>) BigDecimalAccumulator::new,
+                        0.0001,     // magnitudes
+                        0.0001,     // signs
+                        0.0001,     // random cancel
+                        0.0001},    // random values
+                new Object[]{
+                        (Supplier<Accumulator>) KahanAccumulator::new,
+                        0.0001,     // magnitudes
```

Comment:
This margin of error for the test seems extremely large, same goes for Neumaier.

If the relative error is roughly O(εn) then in this case the expected error would be 1E-16 * (4E6) = (4E-10), much smaller than 1E-3.

(The observed error in the test is roughly 2E-16 which is under this expected 4E-10 bound)

Reply:
Yeah ... you caught me. I'll tighten these up.